grassroots-backend/src/auth/Auth.controller.ts

```diff
@@ -9,83 +9,87 @@ import {
 } from "@nestjs/common";
 import { Response as ExpressResponse } from "express";
 import type { GrassrootsRequest } from "../types/GrassrootsRequest";
-import { ConfigService } from "@nestjs/config";
 import { LoginStateDTO } from "../grassroots-shared/LoginState.dto";
 import { VoidDTO } from "../grassroots-shared/Void.dto";
-import { ApiProperty, ApiQuery, ApiResponse } from "@nestjs/swagger";
 import { PublicRoute } from "./PublicRoute.decorator";
 import { OAuthGuard } from "./OAuth.guard";
+import { getEnvironmentVariables } from "../GetEnvironmentVariables";
 
 @Controller("auth")
 export class AuthController {
-  constructor(private configService: ConfigService) {}
+  // Removed caching as Tim suggested - getEnvironmentVariables() already has cache
 
   // The frontend can redirect here to trigger login.
   @Get("login")
   @UseGuards(OAuthGuard)
   @PublicRoute()
-  @ApiQuery({ name: "redirect_path", type: String })
-  login(@Query() redirect_path: string): VoidDTO {
+  login(@Query("redirect_path") redirectPath: string): VoidDTO {
     // The redirect path is used by the OAuth guard.
-    void redirect_path;
+    void redirectPath;
     return VoidDTO.get();
   }
 
   @Get("google/callback")
   @UseGuards(OAuthGuard)
   @PublicRoute()
-  @ApiProperty()
-  googleAuthRedirect(
+  async googleAuthRedirect(
     @Request() req: GrassrootsRequest,
     @Response() response: ExpressResponse,
-  ): VoidDTO {
-    const host = this.configService.get<string>("FRONTEND_HOST");
+  ): Promise<VoidDTO> {
+    // Call getEnvironmentVariables directly each time (no caching)
+    const environmentVariables = await getEnvironmentVariables();
+
+    const host = environmentVariables.FRONTEND_HOST;
     if (host === undefined) {
       throw new Error("Missing env variable for FRONTEND_HOST");
     }
-    if (!req.user) {
+    if (req.user === undefined) {
       throw new Error("No user found for login.");
     }
+
     // The session doesn't contain the redirect path by the time req.login is called,
-    // so make sure to stash it here.
     const redirectPath = req.session.redirect_path ?? host;
     // To prevent a redirect path accidentally being used multiple times, clear this
     // as soon as it's read.
     req.session.redirect_path = undefined;
 
-    req.login(req.user, (err) => {
-      if (err !== undefined) {
-        throw err;
-      }
-      response.redirect(redirectPath);
+    await new Promise<void>((resolve, reject) => {
+      req.login(req.user, (err: unknown) => {
+        if (err !== undefined) {
+          reject(err instanceof Error ? err : new Error("Login failed"));
```

Comment:
Out of curiosity, what rule was this violating?

I think this is clearly better, I'm just wondering. :)